### ISSUE TITLE
Different Entropy Loss Trends Between Original Logs and My Experiments

### ISSUE THREAD

**Issue by hejujie** [agentica-project/rllm#15]
### Description
I noticed different entropy loss trends between the provided wandb logs and my experiments.

1. **In the provided wandb logs:**
   - Entropy loss shows an increasing trend for most experiments (except 24k case)
   - This happens even with a low entropy coefficient (0.001)
![Image](https://github.com/user-attachments/assets/f7e73668-2635-42a9-b9bb-c98d93730ae0)

2. **In my experiments:**
   - Running with the default script from the repository
   - Entropy loss consistently decreases across different response lengths and model sizes
   - This matches the expected behavior in PPO training
![Image](https://github.com/user-attachments/assets/fb932c9d-0d97-49e0-9268-f523042cf40b)

Entropy reflects the exploration capability of LLM during training. A rapid decrease in entropy might lead to premature convergence, preventing the model from achieving the expected performance.

### Questions

1. Is there any explanation for these different trends? 
2. Could this be related to any entropy coefficient configurations that might not be synchronized?

Any insights would be helpful. Thanks!

**Comment by michaelzhiluo** [comment]
Happy to see you caught on to the trends that surprised us too!

In general, we believe for GRPO, only the surrogate loss matters. KL and entropy loss we realized doesn't matter that much.
In our 8K run, KL was incorrectly subtracted from the loss function, leading to exploding KL, but it still worked ;)

Regarding 24K decreasing, I am not sure why, but most likely it has to do with simulated annealing. Restarting a checkpoint from 0 optimizer state can lead to different optimizations for different loss landscapes.

Regarding 8k and 16K entropy increasing, this makes sense, as long chain of thoughts are highly sparse (max tokens ^ max context length), so there is a lot of space to explore as the model improves.

**Comment by hongyi-zhang** [comment]
> Happy to see you caught on to the trends that surprised us too!
> 
> In general, we believe for GRPO, only the surrogate loss matters. KL and entropy loss we realized doesn't matter that much. In our 8K run, KL was incorrectly subtracted from the loss function, leading to exploding KL, but it still worked ;)
> 
> Regarding 24K decreasing, I am not sure why, but most likely it has to do with simulated annealing. Restarting a checkpoint from 0 optimizer state can lead to different optimizations for different loss landscapes.
> 
> Regarding 8k and 16K entropy increasing, this makes sense, as long chain of thoughts are highly sparse (max tokens ^ max context length), so there is a lot of space to explore as the model improves.

Interesting, I was wondering about this as well. Is this related to [this bug fix](https://github.com/volcengine/verl/commit/a65c9157bc0b85b64cd753de19f94e80a11bd871) in verl? Could you clarify in which experiments you subtracted KL from the loss? And have you also tried to rerun the experiments after the fix?

If both KL and entropy terms are subtracted from the loss and with the same coefficient (0.001), you end up adding the cross-entropy w.r.t. the reference model to the maximization objective. Intuitively, this feels like trying to make the policy as unpredictable for the reference model as possible. I now start to wonder if this is a feature rather than a bug ;)

**Comment by michaelzhiluo** [comment]
Only the 8k experiment; in addition, b/c VERL currently has much of its implementation for RLHF, they subtract KL from reward. We removed that too for GRPO.

We haven't tried 8k with the fixed KL loss, but the training should be more robust. It could possibly be a feature and we just simply lucked out!

**Comment by hongyi-zhang** [comment]
> Only the 8k experiment; in addition, b/c VERL currently has much of its implementation for RLHF, they subtract KL from reward. We removed that too for GRPO.
> 
> We haven't tried 8k with the fixed KL loss, but the training should be more robust. It could possibly be a feature and we just simply lucked out!

@michaelzhiluo Following up on this thread: I tried to replicate your original config by setting `actor_rollout_ref.actor.kl_loss_coef=-0.001` (while using `actor_rollout_ref.actor.entropy_coeff=0.001`), but the training trajectory did not resemble your 8k log. Main observations: entropy loss showed a mostly linear decrease in the first 400 steps from 0.27 to 0.05, then slowly descended towards 0.03 in the next 400 steps; KL loss hiked to 0.05 around 400 steps and to 0.11 around 800 steps; critic reward mean plateaued around 200 steps at 0.53 and not yet reaching 0.55 at 800 steps; more importantly, validation accuracy moving average peaked around 220 step at 0.28, then started going downwards to about 0.23 at 800 steps.

I suspect there may be other subtle config/algorithm differences from your original training run? @hejujie how are your experiments going so far?

**Comment by merlinarer** [comment]
I also found the similar trend and cannot reproduce the reported results, as shown in [issue](https://github.com/agentica-project/deepscaler/issues/33). Have tried kl_loss_coef=-0.001 and entropy_coeff=-0.001?

**Comment by JingyangXiang** [comment]
> ### Description
> I noticed different entropy loss trends between the provided wandb logs and my experiments.
> 
> 1. **In the provided wandb logs:**
>    
>    * Entropy loss shows an increasing trend for most experiments (except 24k case)
>    * This happens even with a low entropy coefficient (0.001)
>      ![Image](https://github.com/user-attachments/assets/f7e73668-2635-42a9-b9bb-c98d93730ae0)
> 2. **In my experiments:**
>    
>    * Running with the default script from the repository
>    * Entropy loss consistently decreases across different response lengths and model sizes
>    * This matches the expected behavior in PPO training
>      ![Image](https://github.com/user-attachments/assets/fb932c9d-0d97-49e0-9268-f523042cf40b)
> 
> Entropy reflects the exploration capability of LLM during training. A rapid decrease in entropy might lead to premature convergence, preventing the model from achieving the expected performance.
> 
> ### Questions
> 1. Is there any explanation for these different trends?
> 2. Could this be related to any entropy coefficient configurations that might not be synchronized?
> 
> Any insights would be helpful. Thanks!

When setting `actor_rollout_ref.actor.kl_loss_coef=0.001` and `actor_rollout_ref.actor.entropy_coeff=0.001` and run deepscaler-1.5b-8k experiment, I also meet the same  phenomenon: `Entropy loss consistently decreases across different response lengths and model sizes`.

[![Image](https://github.com/user-attachments/assets/dda0c14a-123a-47e4-b957-fe5223a8cb28)](https://ml.bytedance.net/experiment/tracking/chart?projectId=project_20250214_7d90c34d&snapshotId=view_snapshot_20250219_3a4b2a3a&chartId=tracking-chart-actor-actor/entropy_loss)

**Comment by michaelzhiluo** [comment]
> > Only the 8k experiment; in addition, b/c VERL currently has much of its implementation for RLHF, they subtract KL from reward. We removed that too for GRPO.
> > We haven't tried 8k with the fixed KL loss, but the training should be more robust. It could possibly be a feature and we just simply lucked out!
> 
> [@michaelzhiluo](https://github.com/michaelzhiluo) Following up on this thread: I tried to replicate your original config by setting `actor_rollout_ref.actor.kl_loss_coef=-0.001` (while using `actor_rollout_ref.actor.entropy_coeff=0.001`), but the training trajectory did not resemble your 8k log. Main observations: entropy loss showed a mostly linear decrease in the first 400 steps from 0.27 to 0.05, then slowly descended towards 0.03 in the next 400 steps; KL loss hiked to 0.05 around 400 steps and to 0.11 around 800 steps; critic reward mean plateaued around 200 steps at 0.53 and not yet reaching 0.55 at 800 steps; more importantly, validation accuracy moving average peaked around 220 step at 0.28, then started going downwards to about 0.23 at 800 steps.
> 
> I suspect there may be other subtle config/algorithm differences from your original training run? [@hejujie](https://github.com/hejujie) how are your experiments going so far?



> I also found the similar trend and cannot reproduce the reported results, as shown in [issue](https://github.com/agentica-project/deepscaler/issues/33). [@hongyi-zhang](https://github.com/hongyi-zhang) Have you tried kl_loss_coef=-0.001 and entropy_coeff=-0.001?

This is odd... entropy should increase if kl is set to `-0.001`, not `0.001` as the policy is diverging more.
